### PR TITLE
Only log hearbeat if status changed

### PR DESCRIPTION
**Pull request by arp242**
Right now it will log this every five seconds, which really clobbers my terminal/logs, mostly with fairly useless debug messages:

	 15:27 INFO  producer: Heartbeat
		     num_completed_jobs = 0
		     num_jobs_running   = 0
		     queue              = default
	 15:27 INFO  producer: Heartbeat
		     num_completed_jobs = 0
		     num_jobs_running   = 0
		     queue              = default
	 15:27 INFO  producer: Heartbeat
		     num_completed_jobs = 0
		     num_jobs_running   = 0
		     queue              = default
	 15:27 INFO  producer: Heartbeat
		     num_completed_jobs = 0
		     num_jobs_running   = 0
		     queue              = default

I originally considered changing the InfoContext() to DebugContext(), but doing an Info log only if the status changed seemed like a good middle ground.

### CODE REVIEW

File: producer.go [hunk 1]
```diff
@@ -535,16 +535,25 @@ func (p *producer) dispatchWork(workCtx context.Context, count int, fetchResultC
 func (p *producer) heartbeatLogLoop(ctx context.Context) {
 	ticker := time.NewTicker(5 * time.Second)
 	defer ticker.Stop()
+	type state struct {
```
Comment (brandur) [inline producer.go:538]: I  think the naming on this is just a little too generic. Maybe `jobStatistics`?

File: producer.go [hunk 2]
```diff
@@ -535,16 +535,25 @@ func (p *producer) dispatchWork(workCtx context.Context, count int, fetchResultC
...
+		jobsRan    uint64
+		jobsActive int
+	}
+	prev := state{}
```
Comment (brandur) [inline producer.go:542]: And similarly here too. Maybe:

* `cur` -> `curStats`
* `prev` -> `prevStats`

File: producer.go [hunk 3]
```diff
@@ -535,16 +535,25 @@ func (p *producer) dispatchWork(workCtx context.Context, count int, fetchResultC
...
 			return
 		case <-ticker.C:
-			p.Logger.InfoContext(ctx, p.Name+": Heartbeat",
-				slog.Uint64("num_completed_jobs", p.numJobsRan.Load()),
-				slog.Int("num_jobs_running", int(p.numJobsActive.Load())),
-				slog.String("queue", p.config.Queue),
-			)
+			cur := state{jobsRan: p.numJobsRan.Load(), jobsActive: int(p.numJobsActive.Load())}
+			if cur != prev {
+				p.Logger.InfoContext(ctx, p.Name+": Heartbeat",
```
Comment (brandur) [inline producer.go:550]: Removing the guaranteed tick means this isn't really a heartbeat anymore. Maybe rephrase to?

```suggestion
				p.Logger.InfoContext(ctx, p.Name+": Producer job stats",
```

**Review by brandur — COMMENTED** [review]
Left a couple comments, but WFM.

File: producer.go (hunk 1 above)
Comment (arp242) [reply producer.go]: It's not really "statistics", right? At least, not how I would use the word: It's the state at the time this is run. I changed it to `jobState`, but if you want `jobStatistics` then I can change it to that.

File: producer.go (hunk 3 above)
Comment (arp242) [reply producer.go]: Changed to `p.Logger.InfoContext(ctx, p.Name+": producer job state changed"`

File: producer.go (hunk 1 above)
Comment (brandur) [reply producer.go]: I'm open to alternatives, but the new log message is definitely wrong. Try to zoom out and re-read "producer job state changed" — that sounds like the state of a job is changing. "State" is a common word used in the API for jobs indicating their currently state. e.g. "available", "completed", etc.

Maybe "job counts" or something?

**Comment by ccoVeille** [comment]
I like your changes, but I would like to suggest something.

The heartbeat could be logged on each call with debug level, and the state change is logged with info level.

It would help a lot when a debug is needed.

I'm a bit afraid no log at all could lead to problem

File: producer.go (hunk 1 above)
Comment (arp242) [reply producer.go]: > Maybe "job counts" or something?

Yes, that's better :+1:

File: producer.go [hunk 4]
```diff
@@ -535,16 +535,25 @@ func (p *producer) dispatchWork(workCtx context.Context, count int, fetchResultC
...
 			return
 		case <-ticker.C:
-			p.Logger.InfoContext(ctx, p.Name+": Heartbeat",
-				slog.Uint64("num_completed_jobs", p.numJobsRan.Load()),
-				slog.Int("num_jobs_running", int(p.numJobsActive.Load())),
-				slog.String("queue", p.config.Queue),
-			)
+			curCount := jobCount{ran: p.numJobsRan.Load(), active: int(p.numJobsActive.Load())}
+			if curCount != prevCount {
+				p.Logger.InfoContext(ctx, p.Name+": producer job count changed",
```
Comment (brandur) [inline producer.go:550]: Small, but can you change back to capitalizing the start of sentence? We're pretty consistent on this throughout the rest of the code.

```suggestion
				p.Logger.InfoContext(ctx, p.Name+": Producer job count changed",
```

Comment (arp242) [reply producer.go]: Ah sorry; I'm so used to lower-casing after the `:` I just didn't notice.

**Comment by arp242** [comment]
> I'm a bit afraid no log at all could lead to problem

@ccoVeille  I don't think this will actually log less information? It just won't log if the message will be identical (because none of the log parameters changed).

**Comment by ccoVeille** [comment]
heartbeat will be silent if nothing changed.

That's what your PR is about.

But won't it log nothing for hours if nothing changed. Or maybe I missed something in the conditions.

What will affect the count?

**Comment by brandur** [comment]
@ccoVeille So I think the idea is that while yes, nothing gets logged if nothing is changing, this might be a reasonable compromise for most systems. i.e. Any systems were any jobs are moving through queues at all, you'll see logging basically the same as you did before. But for systems where nothing is happening (like in say dev), we bias towards quieter log outputs.

This seems to make sense to me, although I can imagine an alternative where we should make it a `Debug` line and the client gets even quieter, and those who want a higher output client can either use debug level, or attach their own hooks into the even system.

Thoughts?

File: producer.go [hunk 5]
```diff
@@ -535,16 +535,25 @@ func (p *producer) dispatchWork(workCtx context.Context, count int, fetchResultC
...
 			return
 		case <-ticker.C:
-			p.Logger.InfoContext(ctx, p.Name+": Heartbeat",
-				slog.Uint64("num_completed_jobs", p.numJobsRan.Load()),
-				slog.Int("num_jobs_running", int(p.numJobsActive.Load())),
-				slog.String("queue", p.config.Queue),
-			)
+			curCount := jobCount{ran: p.numJobsRan.Load(), active: int(p.numJobsActive.Load())}
+			if curCount != prevCount {
+				p.Logger.InfoContext(ctx, p.Name+": Producer job count changed",
```
Comment (brandur) [inline producer.go:550]: Can we just make this "counts"? Saying "changed" seems to suggest the log was fired by something changing, which isn't really the case since it's just periodic output.

```suggestion
				p.Logger.InfoContext(ctx, p.Name+": Producer job counts",
```

**Review by brandur — APPROVED** [review]

**Comment by brandur** [comment]
Thanks!

**Comment by ccoVeille** [comment]
I'm fine and agree with what you said.

I might open a PR with what I was thinking about. But my initial comment was a suggestion while the changes involved in the PR were already great.

So no worries
